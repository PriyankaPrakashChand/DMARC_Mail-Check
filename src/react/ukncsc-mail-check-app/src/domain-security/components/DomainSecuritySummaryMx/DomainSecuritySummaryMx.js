import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { Grid, Header, Label } from 'semantic-ui-react';
import kebabCase from 'lodash/kebabCase';
import flatMap from 'lodash/flatMap';
import {
  DomainSecurityRecord,
  DomainSecuritySummaryMessages,
  DomainSecurityTitle,
} from 'domain-security/components';
import { DomainSecurityLocationContext } from 'domain-security/context';
import { MailCheckMessage } from 'common/components';

const shouldShowNoMxRecordsMessage = (type, { pending, records }) =>
  type.startsWith('TLS') &&
  !pending &&
  records &&
  (!records.length || records.some(record => !record.hostname));

const shouldShowRecords = ({ loading, error, records }) =>
  !loading && !error && records;

export default class DomainSecuritySummaryMx extends Component {
  componentDidMount() {
    const {
      domainId,
      fetchDomainSecurityMx,
      getDomainSecurityMx,
      type,
    } = this.props;

    if (type === 'TLS') {
      if (!getDomainSecurityMx(domainId)) {
        fetchDomainSecurityMx(domainId);
      }
    } else {
      this.fetchDomainBasedMx();
    }
  }

  componentDidUpdate() {
    if (this.props.type !== 'TLS') {
      this.fetchDomainBasedMx();
    }
  }

  getIdentifier = () =>
    this.props.type === 'TLS' ? this.props.domainId : this.props.domainName;

  getRecordPrefix = () => (this.props.type === 'DKIM' ? 'Selector : ' : '');

  getNoResultMessage() {
    switch (this.props.type) {
      case 'TLS Certificates':
      case 'TLS':
        return `We're currently checking your MX records. This usually takes less than 15 minutes, so check back soon.`;
      case 'DKIM':
        return `We haven't seen any DKIM selectors for this domain.`;
      default:
        return `We're currently checking your ${
          this.props.type
        } configuration. Please check back later.`;
    }
  }

  fetchDomainBasedMx() {
    const {
      domainName,
      fetchDomainSecurityMx,
      getDomainSecurityMx,
    } = this.props;

    if (domainName && !getDomainSecurityMx(domainName)) {
      fetchDomainSecurityMx(domainName);
    }
  }

  render() {
    const { description, domainId, getDomainSecurityMx, type } = this.props;
    const mx = getDomainSecurityMx(this.getIdentifier()) || {};
    const recordPrefix = this.getRecordPrefix();
    const noResultMesssage = this.getNoResultMessage();

    return (
      <Grid stackable>
        <Grid.Row>
          <Grid.Column width="4">
            <DomainSecurityTitle
              as="h2"
              title={type}
              failures={flatMap(mx.records, r => r.failures || [])}
              warnings={flatMap(mx.records, r => r.warnings || [])}
              inconclusives={flatMap(mx.records, r => r.inconclusives || [])}
              pending={mx.pending}
              loading={mx.loading}
            />
            { type === 'DKIM' && (
              <div>
                <Label color="blue">In Development</Label>
              </div>
            )}
          </Grid.Column>
          <Grid.Column width="8">
            <p>{description}</p>

            {shouldShowRecords(mx) && (
              <React.Fragment>
                {mx.records.filter(record => !!record.hostname).map(record => (
                  <React.Fragment key={record.hostname}>
                    <Header as="h3">
                      {recordPrefix}
                      {record.hostname}
                    </Header>
                    {record.records &&
                      record.records.map(_ => (
                        <DomainSecurityRecord type="DKIM">
                          {_.record}
                        </DomainSecurityRecord>
                      ))}
                    <p>
                      <DomainSecurityLocationContext.Consumer>
                        {location => (
                          <Link
                            to={`/${location}/${domainId}/${kebabCase(
                              type.toLowerCase()
                            )}/${record.id || record.hostname}`}
                          >
                            View more information
                          </Link>
                        )}
                      </DomainSecurityLocationContext.Consumer>
                    </p>

                    <DomainSecuritySummaryMessages
                      type={type}
                      failures={record.failures.length}
                      warnings={record.warnings.length}
                      inconclusives={record.inconclusives.length}
                    />
                  </React.Fragment>
                ))}

                {shouldShowNoMxRecordsMessage(type, mx) && (
                  <MailCheckMessage success>
                    No MX records found.
                  </MailCheckMessage>
                )}
              </React.Fragment>
            )}

            {mx.pending && (
              <MailCheckMessage info>{noResultMesssage}</MailCheckMessage>
            )}

            {mx.error && (
              <MailCheckMessage error>
                There was a problem retrieving the {type} summary for this
                domain: {mx.error.message}
              </MailCheckMessage>
            )}
          </Grid.Column>
        </Grid.Row>
      </Grid>
    );
  }
}
