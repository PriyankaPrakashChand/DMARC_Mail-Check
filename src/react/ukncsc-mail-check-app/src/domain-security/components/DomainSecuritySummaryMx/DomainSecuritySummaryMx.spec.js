import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render } from 'react-testing-library';
import 'dom-testing-library/extend-expect';
import DomainSecuritySummaryMx from './DomainSecuritySummaryMx';

const fetchDomainSecurityMx = jest.fn();
const getDomainSecurityMx = jest.fn();

describe('DomainSecuritySummaryMx', () => {
  let container;

  afterEach(() => jest.resetAllMocks());

  describe('when no records are provided', () => {
    beforeEach(() => {
      getDomainSecurityMx.mockImplementation(() => ({ records: [] }));

      ({ container } = render(
        <MemoryRouter>
          <DomainSecuritySummaryMx
            description=""
            domainId={123}
            fetchDomainSecurityMx={fetchDomainSecurityMx}
            getDomainSecurityMx={getDomainSecurityMx}
            type="TLS"
          />
        </MemoryRouter>
      ));
    });

    test('it should display a success message', () =>
      expect(
        container.getElementsByClassName('ui success message')[0]
      ).toHaveTextContent('No MX records found.'));

    test('it should match the snapshot', () =>
      expect(container).toMatchSnapshot());
  });

  describe('when no failures, warnings or inconclusives are present', () => {
    beforeEach(() => {
      getDomainSecurityMx.mockImplementation(() => ({
        records: [
          {
            id: 1,
            hostname: '1.com',
            failures: [],
            warnings: [],
            inconclusives: [],
          },
        ],
      }));

      ({ container } = render(
        <MemoryRouter>
          <DomainSecuritySummaryMx
            description=""
            domainId={123}
            fetchDomainSecurityMx={fetchDomainSecurityMx}
            getDomainSecurityMx={getDomainSecurityMx}
            type="TLS"
          />
        </MemoryRouter>
      ));
    });

    test('it should display a success message', () =>
      expect(
        container.getElementsByClassName('ui success message')[0]
      ).toHaveTextContent('TLS well configured.'));

    test('it should match the snapshot', () =>
      expect(container).toMatchSnapshot());
  });

  describe('when an error is present', () => {
    beforeEach(() => {
      getDomainSecurityMx.mockImplementation(() => ({
        error: { message: 'Oh noes!' },
      }));

      ({ container } = render(
        <MemoryRouter>
          <DomainSecuritySummaryMx
            description=""
            domainId={123}
            fetchDomainSecurityMx={fetchDomainSecurityMx}
            getDomainSecurityMx={getDomainSecurityMx}
            type="TLS"
          />
        </MemoryRouter>
      ));
    });

    test('it should display a message with the error', () =>
      expect(
        container.getElementsByClassName('ui error message')[0]
      ).toHaveTextContent('oh noes!'));

    test('it should match the snapshot', () =>
      expect(container).toMatchSnapshot());
  });
});
