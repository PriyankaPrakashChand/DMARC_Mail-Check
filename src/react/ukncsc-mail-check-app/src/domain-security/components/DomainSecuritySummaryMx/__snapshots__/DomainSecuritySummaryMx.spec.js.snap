// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomainSecuritySummaryMx when an error is present it should match the snapshot 1`] = `
<div>
  <div
    class="ui stackable grid"
  >
    <div
      class="row"
    >
      <div
        class="four wide column"
      >
        <h2
          class="ui header DomainSecurityTitle--header"
        >
          TLS
        </h2>
        <div
          class="DomainSecurityTitle--icon"
        />
      </div>
      <div
        class="eight wide column"
      >
        <p />
        <div
          style="margin: 0.2em 0em; max-width: 600px;"
          class="ui error message MailCheckMessage--container"
        >
          There was a problem retrieving the 
          TLS
          summary for this domain: 
          Oh noes!
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DomainSecuritySummaryMx when no failures, warnings or inconclusives are present it should match the snapshot 1`] = `
<div>
  <div
    class="ui stackable grid"
  >
    <div
      class="row"
    >
      <div
        class="four wide column"
      >
        <h2
          class="ui header DomainSecurityTitle--header"
        >
          TLS
        </h2>
        <div
          class="DomainSecurityTitle--icon"
        />
      </div>
      <div
        class="eight wide column"
      >
        <p />
        <h3
          class="ui header"
        >
          1.com
        </h3>
        <p>
          <a
            href="/domain-security/123/tls/1"
          >
            View more information
          </a>
        </p>
        <div
          style="margin: 0.2em 0em; max-width: 600px;"
          class="ui success message MailCheckMessage--container"
        >
          TLS
          well configured.
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DomainSecuritySummaryMx when no records are provided it should match the snapshot 1`] = `
<div>
  <div
    class="ui stackable grid"
  >
    <div
      class="row"
    >
      <div
        class="four wide column"
      >
        <h2
          class="ui header DomainSecurityTitle--header"
        >
          TLS
        </h2>
        <div
          class="DomainSecurityTitle--icon"
        />
      </div>
      <div
        class="eight wide column"
      >
        <p />
        <div
          style="margin: 0.2em 0em; max-width: 600px;"
          class="ui success message MailCheckMessage--container"
        >
          No MX records found.
        </div>
      </div>
    </div>
  </div>
</div>
`;
